File: references/autopilot-2.5/sim/matlab/gpsins/kf.m
%	This is a test program to try out an 11-state
% GPS aided INS Kalman filter.  The estimated states
% are the XYZ (NED) positions in local nav frame.  The 
% body velocites u, v, w.  The body attitude in quaternion
% representation.  G is also estimated to account for any
% bias in the vertical accelerometer channel.
%
%	Author: Aaron Kahn
%  copyright 2002
clc;
clear;


global CompSD PosSD VelSD;

% some coversions
deg2rad = pi/180;
rad2deg = 180/pi;


% setup the standard deviations for the R matricies later
CompSD = 0.2;
PosSD = 0.8;
VelSD = 0.2;

% setup the process noise matrix (Q)
Q = diag([0 0 0 0.1 0.1 0.1 0.00001 0.00001 0.00001 0.00001 0.001]);

% setup the state estimate vector [xyz; uvw; Q; g]
Xest = [0; 0; 0; 0; 0; 0; 1; 0; 0; 0; 32.2];
XYZest = Xest([1:3], 1);
UVWest = Xest([4:6], 1);
Qest = Xest([7:10], 1);
Gest = Xest(11,1);

% setup the truth simulation vector and inertias
Xtrue = [0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0];
Ix = 1;
Iy = 1;
Iz = 1;
Ixz = 0;
m = 1;
G = 32.2;

% setup the coverience matrix
P = eye(11);

dt = 0.01;
tf = 90;

% setup some arrays for saving data
REC = zeros(tf/dt, 16);
RECnoise = zeros(tf/dt, 16);
RECest = zeros(tf/dt, 10);


for n = 1:tf/dt
   t = (n-1)*dt;
   
   % simulate the truth
   F = eulerdcm(Xtrue(7:9))*[0;0;-G*m];
   Fx = F(1) + 0.0;
   Fy = F(2) + 0.01*cos(0.3*t);
   Fz = F(3) + 0.0*cos(0.3*t);
   L = 0.001*sin(0.3*t);
   M = 0.00*cos(0.3*t);
   N = 0.00;
   Xtruedot = sixdofe([Fx Fy Fz L M N Ix Iy Iz Ixz m G Xtrue']);
   Xtrue = Xtrue + Xtruedot'*dt;
   Xtrue(7,1) = tohrev(Xtrue(7,1));
   Xtrue(8,1) = tohrev(Xtrue(8,1));
   Xtrue(9,1) = tohrev(Xtrue(9,1));
   AccelTrue = [Xtruedot(1);Xtruedot(2);Xtruedot(3)] + F + eulerwx(Xtrue(4:6))*Xtrue(1:3);
   
   % Add noise to the true values to simulate the data
   XYZtrue = Xtrue([10:12], 1) + 0*randn(3,1);
   UVWtrue = Xtrue([1:3], 1) + 0.0*randn(3,1);
   PQRtrue = Xtrue([4:6], 1) + 0*deg2rad*randn(3,1) + 0.0*[0.1*deg2rad;0;0];
   THETAtrue = Xtrue([7:9], 1);
   Atrue = AccelTrue + 0*randn(3,1);
   
   %%%%%%%%% THE FILTER STARTS HERE %%%%%%%%%
   
   % Propogate the INS states
   [Xest, Xestdot] = propogate_state(XYZest, UVWest, Atrue, PQRtrue, Qest, Gest, dt);
   
   XYZest = Xest([1:3], 1);
   UVWest = Xest([4:6], 1);
   Qest = normq(Xest([7:10], 1)')';
   Gest = Xest(11,1);
   
   % Genarte the A matrix
   A = gena(UVWest, PQRtrue, Qest, Gest);
   
   % Propogate the converience matrix P
   Pdot = A*P + P*A' + Q;
   P = P + Pdot*dt;
   
   % GPS update once per second
   if( mod(t, 1) == 0 )
      [Xest, P] = gpsupdate(XYZtrue, UVWtrue, XYZest, UVWest, Qest, Gest, P);
      
      XYZest = Xest([1:3], 1);
	   UVWest = Xest([4:6], 1);
   	Qest = normq(Xest([7:10], 1)')';
      Gest = Xest(11,1);
      txt = sprintf('GPS Update at Time %f', t);
      disp(txt);
   end
   
   % Compass update 5 times per second
   if( mod(t, 0.2) == 0 )
      [Xest, P] = compassupdate(THETAtrue(3,1), XYZest, UVWest, Qest, Gest, P);
      
      XYZest = Xest([1:3], 1);
	   UVWest = Xest([4:6], 1);
   	Qest = normq(Xest([7:10], 1)')';
      Gest = Xest(11,1);
   end
   
   
   % Record measurement values
   RECnoise(n,1) = t;
   RECnoise(n, 2:4) = UVWtrue';						% body velocity
   RECnoise(n, 5:7) = PQRtrue'*rad2deg;			% pqr true
   RECnoise(n, 8:10) = THETAtrue'*rad2deg;		% true attitude
   RECnoise(n, 11:13) = XYZtrue';					% NED true position
   RECnoise(n, 14:16) = Atrue';				% true body acceleration
   
   
   % Record true values
   XYZtrue = Xtrue([10:12], 1);
   UVWtrue = Xtrue([1:3], 1);
   PQRtrue = Xtrue([4:6], 1);
   THETAtrue = Xtrue([7:9], 1);
   REC(n, 1) = t;
   REC(n, 2:4) = UVWtrue';						% body velocity
   REC(n, 5:7) = PQRtrue'*rad2deg;			% pqr true
   REC(n, 8:10) = THETAtrue'*rad2deg;		% true attitude
   REC(n, 11:13) = XYZtrue';					% NED true position
   REC(n, 14:16) = AccelTrue';				% true body acceleration
   
   % Record estimated values
   RECest(n,1) = t;
   RECest(n, 2:4) = Xest([4:6],1)';			% estimated body velocity
   RECest(n, 5:7) = (quat2euler(Xest([7:10],1)))*rad2deg; %estimated attitude
   RECest(n, 8:10) = Xest([1:3],1)';		% estimated NED position
end

%%%%%% PLOT THE DATA %%%%%%

figure(1);
subplot(3,1,1);
plot(RECnoise(:,1), RECnoise(:,2), 'c:', REC(:,1), REC(:,2), 'b', RECest(:,1), RECest(:,2), 'r');
title('UVW vs. Time');
ylabel('U ft/s');
subplot(3,1,2);
plot(RECnoise(:,1), RECnoise(:,3), 'c:', REC(:,1), REC(:,3), 'b', RECest(:,1), RECest(:,3), 'r');
ylabel('V ft/s');
subplot(3,1,3);
plot(RECnoise(:,1), RECnoise(:,4), 'c:', REC(:,1), REC(:,4), 'b', RECest(:,1), RECest(:,4), 'r');
ylabel('W ft/s');
xlabel('Time sec');

figure(2);
subplot(3,1,1);
plot(RECnoise(:,1), RECnoise(:,5), 'c:', REC(:,1), REC(:,5), 'b');
title('PQR vs. Time');
ylabel('p deg/s');
subplot(3,1,2);
plot(RECnoise(:,1), RECnoise(:,6), 'c:', REC(:,1), REC(:,6), 'b');
ylabel('q deg/s');
subplot(3,1,3);
plot(RECnoise(:,1), RECnoise(:,7), 'c:', REC(:,1), REC(:,7), 'b');
ylabel('r deg/s');
xlabel('Time sec');

figure(3);
subplot(3,1,1);
plot(REC(:,1), REC(:,8), 'b', RECest(:,1), RECest(:,5), 'r');
title('Attitude vs. Time');
ylabel('Roll deg');
subplot(3,1,2);
plot(REC(:,1), REC(:,9), 'b', RECest(:,1), RECest(:,6), 'r');
ylabel('Pitch deg');
subplot(3,1,3);
plot(REC(:,1), REC(:,10), 'b', RECest(:,1), RECest(:,7), 'r');
ylabel('Yaw deg');
xlabel('Time sec');

figure(4);
subplot(3,1,1);
plot(RECnoise(:,1), RECnoise(:,11), 'c:', REC(:,1), REC(:,11), 'b', RECest(:,1), RECest(:,8), 'r');
title('NED vs. Time');
ylabel('North ft');
subplot(3,1,2);
plot(RECnoise(:,1), RECnoise(:,12), 'c:', REC(:,1), REC(:,12), 'b', RECest(:,1), RECest(:,9), 'r');
ylabel('East ft');
subplot(3,1,3);
plot(RECnoise(:,1), RECnoise(:,13), 'c:', REC(:,1), REC(:,13), 'b', RECest(:,1), RECest(:,10), 'r');
ylabel('Down ft');
xlabel('Time sec');

figure(5);
subplot(3,1,1);
plot(RECnoise(:,1), RECnoise(:,14), 'c:', REC(:,1), REC(:,14), 'b');
title('Body Accel vs. Time');
ylabel('X accel ft/s/s');
subplot(3,1,2);
plot(RECnoise(:,1), RECnoise(:,15), 'c:', REC(:,1), REC(:,15), 'b');
ylabel('Y accel ft/s/s');
subplot(3,1,3);
plot(RECnoise(:,1), RECnoise(:,16), 'c:', REC(:,1), REC(:,16), 'b');
ylabel('Z accel ft/s/s');
xlabel('Time sec');


   
   

File: references/autopilot-2.5/sim/src/gpsins/kf.cpp
/* -*- indent-tabs-mode:T; c-basic-offset:8; tab-width:8; -*- vi: set ts=8:
 * $Id: kf.cpp,v 2.0 2002/09/22 02:07:30 tramm Exp $
 *
 * (c) Trammell Hudson
 *
 * Translation of Aaron Kahn's GPS+INS code.
 *
 **************
 *
 *  This file is part of the autopilot simulation package.
 *
 *  Autopilot is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Autopilot is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Autopilot; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */


#include <mat/Matrix.h>
#include <mat/Matrix_Invert.h>
#include <mat/Vector.h>
#include <mat/Quat.h>
#include <mat/Constants.h>
#include <mat/Conversions.h>
#include <cmath>
#include <cstdlib>

using namespace libmat;
using namespace std;

#include "sixdofe.h"
#include "propigate.h"
#include "gena.h"
#include "gps_update.h"
#include "compass_update.h"



/*
 *  Our actual states and Kalman filter variables
 */
static Matrix<11,11>	Q;
static Vector<11>	X_est;
static Vector<3>	XYZ_est;
static Vector<3>	UVW_est;
static Vector<4>	Q_est;
static double		G_est;

static Vector<12>	X_true;
static Vector<4>	inertia;
static double		m = 1.0;
static double		G = 32.2;

static Matrix<11,11>	P;


/*
 *  Routine to reset all the kalman filter variables back to
 * their initial states.
 */
static void
setup_kf( void )
{
	Q.fill( 0.0 );
	Q[ 0][ 0] = 0.0;
	Q[ 1][ 1] = 0.0;
	Q[ 2][ 2] = 0.0;

	Q[ 3][ 3] = 0.1;
	Q[ 4][ 4] = 0.1;
	Q[ 5][ 5] = 0.1;

	Q[ 6][ 6] = 0.00001;
	Q[ 7][ 7] = 0.00001;
	Q[ 8][ 8] = 0.00001;
	Q[ 9][ 9] = 0.00001;

	Q[10][10] = 0.00001;
	Q[11][11] = 0.001;

	// Our position estimate starts at the origin
	XYZ_est[0] = X_est[0] = 0.0;
	XYZ_est[1] = X_est[1] = 0.0;
	XYZ_est[2] = X_est[2] = 0.0;

	// Our velocity estimate starts at zero
	UVW_est[0] = X_est[3] = 0.0;
	UVW_est[1] = X_est[4] = 0.0;
	UVW_est[2] = X_est[5] = 0.0;

	// Our initial orientation is upright
	Q_est[0] = X_est[6] = 1.0;
	Q_est[1] = X_est[7] = 0.0;
	Q_est[2] = X_est[8] = 0.0;
	Q_est[3] = X_est[9] = 0.0;

	// Gravity starts as G.
	G_est = X_est[10] = 32.2;

	// Setup the covariance matrix
	P = eye<11,double>();

	// Zero our true position
	X_true.fill( 0.0 );

	inertia = Vector<4>(
		1.0,	// Ix,
		1.0,	// Iy,
		1.0,	// Iz,
		0.0	// Ixz
	);
}


static double
normalize(
	double		angle
)
{
	if( angle > pi && angle <= 2 * pi )
		return angle - 2 * pi;
	if( angle < -pi && angle > -2 * pi )
		return angle + 2 * pi;
	return angle;
}


/*
 *  Emulates the Matlab randn() call.  Returns a normally distributed
 * random vector of size n.
 */
template<
	const int	n
>
const Vector<n>
randn(
	double		sigma = 1.0
)
{
	Vector<n>	v;

	for( int i=0 ; i<n ; i++ )
		v[i] = 2.0 * sigma * drand48() - sigma;

	return v;
}


int
main( void )
{
	double		dt		= 0.01;
	const double	tf		= 90.0;
	const int	one_hz		= int(1.0 / dt);
	const int	five_hz		= int(0.2 / dt);
	int		step		= 0;

	setup_kf();

	for( double t = 0.0 ; t < tf ; t += dt, step++ )
	{
		// Simulate the truth
		const Vector<3>	true_euler( slice<6,3>( X_true ) );
		const Vector<3> weight( 0, 0, -G * m );
		const Vector<3> F( eulerDC( true_euler ) * weight );
		const Vector<3> next_F( F + Vector<3>(
			0.01,
			0.0 * cos( 0.3 * t ), 
			0.0 * sin( 0.3 * t )
		));

		const Vector<3> LMN(
			0.01 * sin( 0.3 * t ),
			0.01 * cos( 0.3 * t ),
			0.001
		);

		const Vector<12> Xtrue_dot( sixdofe(
			next_F,
			LMN,
			inertia,
			m,
			G,
			X_true
		) );


		// Update our true position
		X_true += Xtrue_dot * dt;
		X_true[6] = normalize( X_true[6] );
		X_true[7] = normalize( X_true[7] );
		X_true[8] = normalize( X_true[8] );


		// Add noise into true values to simulate the data
		const Vector<3> accel_true(
			slice<0,3>( Xtrue_dot )
			+ F
			+ eulerWx( slice<3,3>(X_true) ) * slice<0,3>(X_true)
			+ randn<3>( 2.5 )
		);

		const double DEG = 2 * C_DEG2RAD;

		const Vector<3> XYZ_true( slice<9,3>(X_true) + randn<3>() );
		const Vector<3> UVW_true( slice<0,3>(X_true) + randn<3>(0.01) );
		const Vector<3> PQR_true( slice<3,3>(X_true) + randn<3>(DEG) );
		const Vector<3> THETA_true( slice<6,3>(X_true) + randn<3>(2.5) );

		// Start the actual filter now
		Vector<11> Xest_dot;

		propigate_state(
			// Outputs
			X_est,
			Xest_dot,

			// Inputs
			XYZ_est,
			UVW_est,
			accel_true,
			PQR_true,
			Q_est,
			G_est,
			dt
		);

		XYZ_est = slice<0,3>( X_est );
		UVW_est = slice<3,3>( X_est );
		Q_est = slice<6,4>( X_est );
		Q_est.norm_self();
		G_est = X_est[10];


		// Generate the A matrix
		const Matrix<11,11> A( gena(
			UVW_est,
			PQR_true,
			Q_est,
			G_est
		) );


		// Propigate the covariance matrix P
		P += (A*P + P * A.transpose() + Q) * dt;

		// GPS update once per second
		if( step % one_hz == 0 )
		{
			gps_update(
				X_est,
				P,
				XYZ_true,
				UVW_true,
				XYZ_est,
				UVW_est,
				Q_est,
				G_est
			);

			XYZ_est = slice<0,3>( X_est );
			UVW_est = slice<3,3>( X_est );
			Q_est = slice<6,4>( X_est );
			Q_est.norm_self();
			G_est = X_est[10];
		}

		// Compass update five times per second
		if( step % five_hz == 0 )
		{
			compass_update(
				X_est,
				P,
				THETA_true[2],
				XYZ_est,
				UVW_est,
				Q_est,
				G_est
			);

			XYZ_est = slice<0,3>( X_est );
			UVW_est = slice<3,3>( X_est );
			Q_est = slice<6,4>( X_est );
			Q_est.norm_self();
			G_est = X_est[10];
		}

		cout
			<< XYZ_est[0] << " "
			<< XYZ_true[0] << " "
			<< X_true[9] << endl;
	}
}
